docs/tp/04.md

```diff
@@ -0,0 +1,126 @@
+* Lorsque vous avez une grande quantité de données et que vous souhaitez transférer la propriété, mais assurez-vous que les données ne seront pas copiées lorsque vous le ferez
+* Lorsque vous voulez posséder une valeur et que vous vous souciez seulement qu'il s'agisse d'un type qui implémente un trait particulier plutôt que d'être d'un type spécifique
+
+```rust
+fn main() {
+    let b = Box::new(5);
+    println!("b = {}", b);
+}
+```
+
+Nous définissons la variable b comme ayant la valeur d'une `Box` qui pointe vers la valeur 5, qui est allouée sur le **tas**. Ce programme imprimera `b = 5`; dans ce cas, nous pouvons accéder aux données dans la boîte de la même manière que nous le ferions si ces données étaient sur la **pile**. Comme toute valeur possédée, lorsqu'une boîte sort de la portée, comme le fait b à la fin de main, elle sera désallouée. La désallocation se produit à la fois pour la boîte (stockée sur la pile) et les données vers lesquelles elle pointe (stockées sur le tas).
+
+## Processus
+Lancer une application signifie allouer des ressources système (processeur, mémoire, périphériques d'entrée / sortie) pour exécuter l'application. Une application qui s'exécute, c'est-à-dire utilise des ressources système pour exécuter du code et traiter des données, est appelée un **processus**. Lorsque nous démarrons une application, un processus est créé; lorsque nous arrêtons l'application, ou lorsqu'elle termine son exécution, nous entendons la fin de l'exécution du processus.
+
+Le processus démarre à partir d'un fichier exécutable contenant le code de l'application (instructions) et les données. Le fichier exécutable est également appelé **image du processus** . Le fichier exécutable est un programme. Nous disons que le processus est un programme en cours d'exécution. 
+
+### Identifier un processus
+``PID`` ( Process Id ) est l'attribut essentiel du processus, un index qui identifie le processus au niveau du système. Un processus est identifié par ``PID`` et non par nom d'exécutable (CMD). Nous pouvons créer plusieurs processus à partir du même exécutable, chaque processus ayant son propre ``PID``.
+
+### Fork
+La fonction ``fork()`` crée un nouveau processus enfant dupliquant le processus parent. Après avoir appelé l'appel système fork (avec succès), deux processus seront créés qui sont identiques à l'exception de leur pid et de la valeur de retour de cette fonction.
+
+```rust
+use nix::sys::wait::waitpid;
+use nix::sys::unistd::{fork, ForkResult, write};
+ 
+match unsafe{fork()} {
+   Ok(ForkResult::Parent { child, .. }) => {
+       println!("Continuing execution in parent process, new child has pid: {}", child);
+       waitpid(child, None).unwrap();
+   }
+   Ok(ForkResult::Child) => {
+       // Unsafe to use `println!` (or `unwrap`) here. See Safety.
```

Comment:
Why?

Reply:
Got it, this is valid only in multithreaded software, which is not the case here. You can use `println!`.